packages/core/admin/server/src/services/homepage.ts

```diff
@@ -0,0 +1,131 @@
+import type { Core, UID } from '@strapi/types';
+import { contentTypes } from '@strapi/utils';
+import type { GetRecentDocuments } from '../../../shared/contracts/homepage';
+
+const createHomepageService = ({ strapi }: { strapi: Core.Strapi }) => {
+  const MAX_DOCUMENTS = 4;
+
+  const metadataService = strapi.plugin('content-manager').service('document-metadata');
+  const permissionService = strapi.admin.services.permission as typeof import('./permission');
+
+  /**
+   * Don't use the strapi.store util because we need to make
+   * more precise queries than exact key matches, in order to make as few queries as possible.
+   */
+  const coreStore = strapi.db.query('strapi::core-store');
+
+  return {
+    async getRecentUpdates(): Promise<GetRecentDocuments.Response['data']> {
+      // Get all the content types the user has permissions to read
+      const readPermissions = await permissionService.findMany({
+        where: {
+          role: { users: { id: strapi.requestContext.get()?.state?.user.id } },
+          action: 'plugin::content-manager.explorer.read',
+        },
+      });
+      const allowedContentTypeNames = readPermissions
+        .map((permission) => permission.subject)
+        .filter(Boolean) as UID.ContentType[];
+
+      // Fetch the configuration for each content type in a single query
+      const rawConfigurations = await coreStore.findMany({
+        where: {
+          key: {
+            $in: allowedContentTypeNames.map(
+              (contentType) => `plugin_content_manager_configuration_content_types::${contentType}`
+            ),
+          },
+        },
+      });
+      const configurations = rawConfigurations.map((rawConfiguration) => {
+        return JSON.parse(rawConfiguration.value);
+      });
+
+      const recentDocuments = await Promise.all(
+        allowedContentTypeNames.map(async (contentTypeName) => {
+          const configuration = configurations.find((config) => config.uid === contentTypeName);
+          const contentType = strapi.contentType(contentTypeName);
+          const fields = ['documentId', 'updatedAt'];
+
+          // Add fields required to get the status if D&P is enabled
+          const hasDraftAndPublish = contentTypes.hasDraftAndPublish(contentType);
+          if (hasDraftAndPublish) {
+            fields.push('publishedAt');
+          }
+
+          // Only add the main field if it's defined
+          const { mainField } = configuration.settings;
+          if (mainField) {
+            fields.push(mainField);
+          }
+
+          // Only add locale if it's localized
+          const isLocalized = (contentType.pluginOptions?.i18n as any)?.localized;
+          if (isLocalized) {
+            fields.push('locale');
+          }
+
+          const documents = await strapi.documents(contentTypeName).findMany({
+            limit: MAX_DOCUMENTS,
+            sort: 'updatedAt:desc',
```

Comment:
Does this really do anything for us or how come you also have to sort updatedAt on 101?

Reply:
This makes sure we only fetch the most recent stuff for each content type. Line 101 compares a blend of all content types to only keep the most recent stuff overall